### ISSUE TITLE
Caching of provisioned tools

### ISSUE THREAD

**Issue by Vampire** [pbrisbin/setup-tool-action#83]
The description of your action suggests that the provisioned tool will be cached.
But this is not the case, it will be re-downloaded on every workflow run.

Unfortunately the name and description of `@actions/tool-cache` is pretty misleading.
Just using it does not put anything to the GHA cache.
Only to the tool-cache directory or taking it from there if it is available,
so you can switch back and forth between different version, or have some tools already present in the hosted runners or also custom tools if you use custom runners.
Only between steps of the same workflow this "cache" for a newly downloaded tool could work.

Is this the intended behavior of your action?
If not, this is bug report.
If yes, this is a feature request. :-)

To put the downloaded stuff to the GHA cache, you need to additionally use `@actions/cache`.

You can an example algorithm for that in my `setup-wsl` action, it is implemented in Kotlin though: https://github.com/Vampire/setup-wsl/blob/master/src/main/kotlin/net/kautler/github/action/setup_wsl/SetupWsl.kt
Basically I have a `use-cache` input that you can set to `true` or `false`.
If it is not set `isFeatureAvailable` of `@actions/cache` is used to determine and if `true` it is also used to verify the cache is available.
And then if the cache should be used I use `restoreCache` and `saveCache` accordingly to restore a cache or save it if there was no cached version available.

**Comment by pbrisbin** [comment]
> The description of your action suggests that the provisioned tool will be cached

Whoops, I've fixed this. I was indeed thrown off by `tool-cache` when I first set up this repo, and didn't update the description after actually implementing and understanding the situation.

> Is this the intended behavior of your action?

Yes. My belief is that, given the primary use-case is downloading a binary, there would be little time saved, if any, by downloading and extracting that binary instead.

So I'm unlikely to implement this myself at this time. I would accept a PR that implements, if it also linked to some example workflow runs that showed the caching indeed saves time. Maybe you have some workflows that use `setup-wsl` where we can compare cache hit-vs-miss runs to see if it's worth the complexity to implement.

**Comment by Vampire** [comment]
Well, it's not only about performance, also about reliability.
I've run a certain workflow multiple times in parallel to reproduce a failure situation.
At that time I just used `wget` to download the tool.
This resulted in some of the tries fail due to download error.
Maybe it was an API ban for triggering the download too often in two little time, or it was a server outage, I don't know.
That was when I searched for an action that does the downloading and also caches the downloaded file / unpacked tool.
If your action wouldn't have stated it does caching, I probably wouldn't have tried it after all.

It's not too likely I'll contribute the implementation.
Initially I've written `setup-wsl` in typescript.
It was a real PITA like every time when I have to touch JS or TS.
So I rewrote it in Kotlin/JS which then was real fun and working much better. :-)

**Comment by Vampire** [comment]
Well, the work-around is easy enough, adding a `actions/cache` action in front. :-)
I just fear that I can then also switch back to using `wget`, as for this occasion I think your action then does not add any value.
But good to have the improvements and typings for the next time I need something like this. :-)
